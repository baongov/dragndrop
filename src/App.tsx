import React, { useState } from "react";
import styled from "styled-components";
import _get from "lodash/get";

import {
  generateColorMatrix,
  isEqual,
  swapBlock,
  stringToPosition,
  Position,
  Block,
} from "./utils";
import { BLOCK_SIZE } from "./config";

const EMPTY_POINT: Position = [-1, -1];

interface ColorBlockProps {
  h: number;
  s: number;
  l: number;
  x: number;
  y: number;
  isSelected: boolean;
  isDraggerOn: boolean;
}

const getBlockStyleByState = (props: ColorBlockProps) => {
  const { isDraggerOn, isSelected } = props;
  if (isDraggerOn) {
    return `
      border: 3px solid yellow;
      z-index: 200;
    `;
  }

  if (isSelected) {
    return `
      border: 3px solid black;
      z-index: 150;
    `;
  }

  return `
    border: 1px solid white;
    z-index: 0;
  `;
};

const ColorBlock = styled.div`
  box-sizing: border-box;
  width: ${BLOCK_SIZE}px;
  height: ${BLOCK_SIZE}px;
  background-color: ${({ h, s, l }: ColorBlockProps) =>
    `hsl(${h}, ${s}%, ${l}%)`};
  top: ${({ x }) => x}px;
  left: ${({ y }) => y}px;
  position: fixed;
  ${(props) => getBlockStyleByState(props)}

  &:hover {
    border: 3px solid brown;
    z-index: 100;
  }
`;

const Container = styled.div`
  padding-left: 100px;
`;

interface AppProps {
  size: number;
}
const App = ({ size }: AppProps) => {
  const [board, setBoard] = useState<Block[][]>(
    generateColorMatrix(size, [100, 100])
  );
  const [selectedBlock, setSelectedBlock] = useState<Position>(EMPTY_POINT);

  // drap and drop
  const [draggedBlock, setDraggedBlock] = useState<Position>(EMPTY_POINT);
  const [hoveredBlock, setHoveredBlock] = useState<Position>(EMPTY_POINT);

  const onDragStart = (id: Position) => {
    setDraggedBlock(id);
  };

  const onDragOver = (event: any) => {
    event.preventDefault();
  };

  const onDrop = () => {
    setBoard(swapBlock(board, draggedBlock, hoveredBlock));
  };

  const onDragEnter = (event: any) => {
    const position: string = _get(event, "currentTarget.dataset.position");
    setHoveredBlock(stringToPosition(position));
  };

  const onDragEnd = () => {
    setDraggedBlock(EMPTY_POINT);
    setHoveredBlock(EMPTY_POINT);
  };

  return (
    <Container className="App">
      <div>Selected position: {JSON.stringify(selectedBlock)}</div>
      <div>Hovered position: {JSON.stringify(hoveredBlock)}</div>
      <div>Dragged position: {JSON.stringify(draggedBlock)}</div>
      <div>
        {board.map((row, i) => (
          <div key={i}>
            {row.map((column, j) => {
              const id: Position = [i, j];
              const { color, position } = column;
              const [h, s, l] = color;
              const [x, y] = position;
              const isSelected = isEqual(id, selectedBlock);

              return (
                <ColorBlock
                  key={j}
                  // drap and drop
                  draggable
                  data-position={id}
                  onDragStart={() => onDragStart(id)}
                  onDragOver={onDragOver}
                  onDrop={onDrop}
                  onDragEnter={onDragEnter}
                  onDragEnd={onDragEnd}
                  // style
                  h={h}
                  s={s}
                  l={l}
                  x={x}
                  y={y}
                  // others
                  isSelected={isSelected}
                  isDraggerOn={isEqual(id, hoveredBlock)}
                  onClick={() => setSelectedBlock(id)}
                />
              );
            })}
          </div>
        ))}
      </div>
    </Container>
  );
};

export default App;
