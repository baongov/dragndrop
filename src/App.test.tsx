import React from "react";
import { shallow } from "enzyme";
import App from "./App";

describe("renders learn react link", () => {
  it("Renders link to Google", () => {
    const app = shallow(<App size={2} />);
    expect(app).toMatchSnapshot();
  });
});
