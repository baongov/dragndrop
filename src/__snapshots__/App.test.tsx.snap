// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders learn react link Renders link to Google 1`] = `
<styled.div
  className="App"
>
  <div>
    Selected position: 
    [-1,-1]
  </div>
  <div>
    Hovered position: 
    [-1,-1]
  </div>
  <div>
    Dragged position: 
    [-1,-1]
  </div>
  <div>
    <div
      key="0"
    >
      <styled.div
        data-position={
          Array [
            0,
            0,
          ]
        }
        draggable={true}
        h={0}
        isDraggerOn={false}
        isSelected={false}
        key="0"
        l={10}
        onClick={[Function]}
        onDragEnd={[Function]}
        onDragEnter={[Function]}
        onDragOver={[Function]}
        onDragStart={[Function]}
        onDrop={[Function]}
        s={80}
        x={100}
        y={100}
      />
      <styled.div
        data-position={
          Array [
            0,
            1,
          ]
        }
        draggable={true}
        h={0}
        isDraggerOn={false}
        isSelected={false}
        key="1"
        l={90}
        onClick={[Function]}
        onDragEnd={[Function]}
        onDragEnter={[Function]}
        onDragOver={[Function]}
        onDragStart={[Function]}
        onDrop={[Function]}
        s={80}
        x={100}
        y={150}
      />
    </div>
    <div
      key="1"
    >
      <styled.div
        data-position={
          Array [
            1,
            0,
          ]
        }
        draggable={true}
        h={320}
        isDraggerOn={false}
        isSelected={false}
        key="0"
        l={10}
        onClick={[Function]}
        onDragEnd={[Function]}
        onDragEnter={[Function]}
        onDragOver={[Function]}
        onDragStart={[Function]}
        onDrop={[Function]}
        s={80}
        x={150}
        y={100}
      />
      <styled.div
        data-position={
          Array [
            1,
            1,
          ]
        }
        draggable={true}
        h={320}
        isDraggerOn={false}
        isSelected={false}
        key="1"
        l={90}
        onClick={[Function]}
        onDragEnd={[Function]}
        onDragEnter={[Function]}
        onDragOver={[Function]}
        onDragStart={[Function]}
        onDrop={[Function]}
        s={80}
        x={150}
        y={150}
      />
    </div>
  </div>
</styled.div>
`;
